### ISSUE TITLE
Build error on Windows - Related to missing "zconf.h"

### ISSUE THREAD

**Issue by Elaws** [ahrm/sioyek#143]
Hello,

I get several errors when trying to build on Windows, all related to a missing `zconf.h`. 

I have done the following :

1. Install VS 2019
2. Install component C++ for desktop
3. Install Qt 15
4. Cloned the repo using `git clone --recursive`
5. Launched the `build_windows.bat`

Here are some errors I get along the build :

```
 E:\...\Documents\Mes Documents\Code\sioyek\zlib\zlib.h(34,10): fatal error C1083: Cannot open include file: 'zconf.h': No such file or directory [E:\...\Documents\Mes Documents\Code\sioyek\sioyek.vcxproj]
```

```
 E:\...\Documents\Mes Documents\Code\sioyek\mupdf\thirdparty\zlib\zlib.h(34,10): fatal error C1083: Cannot open include file: 'zconf.h': No such file or directory [E:\...\Documents\Mes Documents\Code\sioyek\mupdf\platform\win32\libthirdparty.vcxproj]
```

```
 E:\...\Documents\Mes Documents\Code\sioyek\mupdf\thirdparty\zlib\zlib.h(34,10): fatal error C1083: Cannot open include file: 'zconf.h': No such file or directory [E:\...\Documents\Mes Documents\Code\sioyek\mupdf\platform\win32\libextract.vcxproj]
```

I can add more details if needed. Thanks for your help.

**Comment by ahrm** [comment]
Did you use Visual Studio command prompt?

**Comment by Elaws** [comment]
I used `View > Terminal' inside Visual Studio Community 2019. It is written `Developer PowerShell`.

**Comment by ahrm** [comment]
Hmmm I am not entirely sure. Here is the github action that build the windows version:
```
  build-windows:

    runs-on: windows-latest

    steps:
    - uses: actions/checkout@v2
      with:
            submodules: 'recursive'
    
    - name: Install Qt
      uses: jurplel/install-qt-action@v2

    - name: Add MSBuild to PATH
      uses: microsoft/setup-msbuild@v1.0.2
    - name: Add msvc-dev-cmd
      uses: ilammy/msvc-dev-cmd@v1
      
      
    - name: Build Sioyek
      working-directory: ${{env.GITHUB_WORKSPACE}}
      run: .\build_windows.bat non_portable
      
    - name: upload windows artifact
      uses: actions/upload-artifact@v1
      with:
        name: uploads
        path: sioyek-release-windows.zip

```
Maybe it can help.

**Comment by Elaws** [comment]
Thanks. Ok, no more problems with "zconf.h" (I cloned the folder in the wrong place). 

I can get much further in the build, but wow I get errors at :

```msbuild -maxcpucount sioyek.vcxproj /property:Configuration=Release```

I get hundreds of warnings of this kind :

```
   1>E:\...\Documents\Mes Documents\EssaiBuild\sioyek\pdf_viewer\database.cpp(194,81): warning C4100: 'col_name' : paramètre formel non référencé [E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxproj]
     1>E:\...\Documents\Mes Documents\EssaiBuild\sioyek\pdf_viewer\database.cpp(194,55): warning C4100: 'argc' : paramètre formel non référencé [E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxproj]
     1>E:\...\Documents\Mes Documents\EssaiBuild\sioyek\pdf_viewer\database.cpp(853,11): warning C4189: 'prev_database' : variable locale initialisée mais non référencée [E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxpr
       oj]
     1>E:\...\Documents\Mes Documents\EssaiBuild\sioyek\pdf_viewer\database.cpp(1003,20): warning C4018: '<' : incompatibilité signed/unsigned [E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxproj]
     1>E:\...\Documents\Mes Documents\EssaiBuild\sioyek\pdf_viewer\database.cpp(1165,20): warning C4018: '<' : incompatibilité signed/unsigned [E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxproj]
     1>E:\...\Documents\Mes Documents\EssaiBuild\sioyek\pdf_viewer\database.cpp(1167,9): warning C4018: '<' : incompatibilité signed/unsigned [E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxproj]
```

And finally this error message :

```
 LINK : fatal error LNK1181: impossible d'ouvrir le fichier en entrée 'libmupdf.lib' [E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxproj]
```

If it helps, I have a `libmupdf.lib` in `\sioyek\mupdf\platform\win32\Release`, and not in `\sioyek\mupdf\platform\win32\x64\Release` as it seems to be expected by `pdf_viewer_build_config.pro`.

I am using `Developer command prompt for VS 2019` standalone (= without opening VS Studio).

**Comment by ahrm** [comment]
The warnings are not important. Maybe manually create the `x64` folder and copy the relevant files there?

**Comment by Elaws** [comment]
Ok, so I moved `\mupdf\platform\win32\Release` to `\mupdf\platform\win32\x64\Release`, and went further again.

But now I get a warning before 1118 errors. The warning is :

```LINK : warning LNK4098: conflit entre la bibliothèque par défaut 'LIBCMT' et les autres bibliothèques ; utilisez /NODEFAULTLIB:library [E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxproj]```

The errors are of this kind : 

```
"E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxproj" (cible par défaut) (1) ->
       (Link cible) ->
         RunGuard.obj : error LNK2001: symbole externe non résolu "__declspec(dllimport) public: __thiscall QMessageLogger::QMessageLogger(char const *,int,char const *)" (__imp_??0QMessageLogger@@QAE@PBDH0@Z) [E:\...\Documents\Mes Do
       cuments\EssaiBuild\sioyek\sioyek.vcxproj]
         RunGuard.obj : error LNK2001: symbole externe non résolu "__declspec(dllimport) public: class QDebug __thiscall QMessageLogger::debug(void)const " (__imp_?debug@QMessageLogger@@QBE?AVQDebug@@XZ) [E:\...\Documents\Mes Document
       s\EssaiBuild\sioyek\sioyek.vcxproj]
         RunGuard.obj : error LNK2001: symbole externe non résolu "__declspec(dllimport) public: class QDebug __thiscall QMessageLogger::critical(void)const " (__imp_?critical@QMessageLogger@@QBE?AVQDebug@@XZ) [E:\...\Documents\Mes Do
       cuments\EssaiBuild\sioyek\sioyek.vcxproj]
         utils.obj : error LNK2001: symbole externe non résolu "__declspec(dllimport) public: __thiscall QMetaObject::Connection::~Connection(void)" (__imp_??1Connection@QMetaObject@@QAE@XZ) [E:\...\Documents\Mes Documents\EssaiBuild\
       sioyek\sioyek.vcxproj]
         RunGuard.obj : error LNK2001: symbole externe non résolu "__declspec(dllimport) public: __thiscall QMetaObject::Connection::~Connection(void)" (__imp_??1Connection@QMetaObject@@QAE@XZ) [E:\...\Documents\Mes Documents\EssaiBui
       ld\sioyek\sioyek.vcxproj]
```

**Comment by ahrm** [comment]
You should use 64 bit developer command prompt in order to build sioyek.  You can open it by opening start menu in windows and searching for `x64 Native Tools Command Prompt`.

**Comment by Elaws** [comment]
Getting there: 1118 -> 8 errors !

I still have a few unresolved external errors :

```
"E:\...\Documents\Mes Documents\EssaiBuild\sioyek\mupdf\platform\win32\mupdf.sln" (cible par défaut) (1) ->
"E:\...\Documents\Mes Documents\EssaiBuild\sioyek\mupdf\platform\win32\mupdf-curl.vcxproj" (cible par défaut) (15) ->
(Link cible) ->
  curl_stream.obj : error LNK2001: symbole externe non résolu __imp_curl_easy_perform [E:\...\Documents\Mes Documents
\EssaiBuild\sioyek\mupdf\platform\win32\mupdf-curl.vcxproj]
  curl_stream.obj : error LNK2001: symbole externe non résolu __imp_curl_global_init [E:\...\Documents\Mes Documents\
EssaiBuild\sioyek\mupdf\platform\win32\mupdf-curl.vcxproj]
  curl_stream.obj : error LNK2001: symbole externe non résolu __imp_curl_easy_init [E:\...\Documents\Mes Documents\Es
saiBuild\sioyek\mupdf\platform\win32\mupdf-curl.vcxproj]
  curl_stream.obj : error LNK2001: symbole externe non résolu __imp_curl_easy_strerror [E:\...\Documents\Mes Document
s\EssaiBuild\sioyek\mupdf\platform\win32\mupdf-curl.vcxproj]
  curl_stream.obj : error LNK2001: symbole externe non résolu __imp_curl_easy_cleanup [E:\...\Documents\Mes Documents
\EssaiBuild\sioyek\mupdf\platform\win32\mupdf-curl.vcxproj]
  curl_stream.obj : error LNK2001: symbole externe non résolu __imp_curl_easy_setopt [E:\...\Documents\Mes Documents\
EssaiBuild\sioyek\mupdf\platform\win32\mupdf-curl.vcxproj]
  x64\Release\mupdf-curl.exe : fatal error LNK1120: 6 externes non résolus [E:\...\Documents\Mes Documents\EssaiBuild
\sioyek\mupdf\platform\win32\mupdf-curl.vcxproj]
```

```
 "E:\...\Documents\Mes Documents\EssaiBuild\sioyek\sioyek.vcxproj" (cible par défaut) (1) ->
       (Link cible) ->
         synctex_parser.obj : error LNK2001: symbole externe non résolu gzread [E:\...\Documents\Mes Documents\EssaiB
       uild\sioyek\sioyek.vcxproj]
         synctex_parser.obj : error LNK2001: symbole externe non résolu gzprintf [E:\...\Documents\Mes Documents\Essa
       iBuild\sioyek\sioyek.vcxproj]
         synctex_parser.obj : error LNK2001: symbole externe non résolu gzclose [E:\...\Documents\Mes Documents\Essai
       Build\sioyek\sioyek.vcxproj]
         synctex_parser.obj : error LNK2001: symbole externe non résolu gzerror [E:\...\Documents\Mes Documents\Essai
       Build\sioyek\sioyek.vcxproj]
         synctex_parser.obj : error LNK2001: symbole externe non résolu gzopen [E:\...\Documents\Mes Documents\EssaiB
       uild\sioyek\sioyek.vcxproj]
         synctex_parser.obj : error LNK2001: symbole externe non résolu gzseek [E:\...\Documents\Mes Documents\EssaiB
       uild\sioyek\sioyek.vcxproj]
         synctex_parser.obj : error LNK2001: symbole externe non résolu gztell [E:\...\Documents\Mes Documents\EssaiB
       uild\sioyek\sioyek.vcxproj]
         release\\sioyek.exe : fatal error LNK1120: 7 externes non résolus [E:\...\Documents\Mes Documents\EssaiBuild
       \sioyek\sioyek.vcxproj]
```

**Comment by Elaws** [comment]
I'm wondering if I'm using the proper `msbuild` ? When I write `where msbuild`, I get two results :

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\MSBuild.exe
C:\Windows\Microsoft.NET\Framework64\v4.0.30319\MSBuild.exe
```

Could this explain the previous errors ?

**Comment by ahrm** [comment]
You should be using the first one. I am not sure if this is the cause of the error.

**Comment by Elaws** [comment]
So, I started from scratch again (from cloning the repo), this time everything is done with `x64 Native Tools Command Prompt for VS 2019`.

(Please note that in the `build_windows.bat`, I replaced all the `cp` by `copy` (`cp` led to errors).)

I keep getting the errors mentioned [here](https://github.com/ahrm/sioyek/issues/143#issuecomment-1050126797). However, it builds successfully and `sioyek-release-windows` folder is created.

When launching `sioyek.exe`, I had a blank page (100% white). It turned out that `sioyek-release-windows` folder was missing the `shaders` folder. Indeed, on line in the `build_windows.bat` did not execute propely :

```
...\sioyek>copy -r pdf_viewer\shaders sioyek-release-windows\shaders
Specified file can't be found.
```

But `pdf_viewer\shaders` folder does exist. So I just copy pasted that folder into  `sioyek-release-windows` and now the PDF are displayed properly !

However, the bottom information bar is missing :

![image](https://user-images.githubusercontent.com/52013479/155606968-c2eb2609-6fab-49b0-bf40-c928c63c0131.png)

Is this normal in this version ?

**Comment by ahrm** [comment]
> Is this normal that there is no information bar (where page number is usually displayed) ?

No, this was a bug in latest commit. Should be fixed in 250359a866d39a5f67ce6b3644b3ed78e5899958.

> Why shaders folder is not copied properly ?

Because copy command doesn't have the `-r` option which is used to copy folders.

>Is it OK that I replace cp by copy in the .bat ?

No, you should also replace the command line options (such as `-r`) with the correct alternatives.

>Why does it build successfully even if I have 15 unresolved references errors ?

I assume the errors where related to some components of `mupdf` which are not used in sioyek.

**Comment by Elaws** [comment]
Thank a lot @ahrm, I am now successfully able to compile and run your code !  **I consider this issue solved.** Thanks again for your help and patience regarding these beginner's issues.

I modified my sources with 250359a866d39a5f67ce6b3644b3ed78e5899958, compiled again, and confirm the status bar is there ! 

---
To whom it may help, please find below details to help you build on Windows :

# Steps for proper build on Windows 

- Install *Qt 5* and *VS development with C++* as shown [here](https://github.com/ahrm/sioyek/issues/143#issuecomment-1049710090).
- Add `...\Qt\5.15.2\msvc2019_64\bin` and `...\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin` (replace `...\` with your proper path) to environment variable *path*.
- Do all the commands in *x64 Native Tools Command Prompt for VS 2019* (`git clone --recursive https://github.com/ahrm/sioyek`, `cd sioyek` and `build_windows.bat portable` or `build_windows.bat non-portable`).
- Do the proper replacements in *build_windows.bat* (see below). 
- At first, the error I had with missing *zconf.h* was due to the fact that I opened that repo inside Visual Studio 2019. Something weird happened (with cmake ?) where my *zconf.h* was automatically replaced with *zconf.h.cmakein* (hence *zconf.h* missing). So I just cloned *Sioyek* repo in a local folder on my computer **without** opening it inside VS Studio.  

## Replacements in *build_windows.bat* :

- `rm -r sioyek-release-windows 2> NUL` by `rmdir /S sioyek-release-windows`
- `cp -r pdf_viewer\shaders sioyek-release-windows\shaders\` by `xcopy /E /I pdf_viewer\shaders sioyek-release-windows\shaders\` (**last \ after shaders is important !**).
- Remaining `cp` by `copy`.
